### ISSUE TITLE
EPA failure - another example of an EPA exception when computing signed distance

### ISSUE THREAD

**Issue by SeanCurtis-TRI** [flexible-collision-library/fcl#493]
In the wild, another signed distance query produced an EPA failure. Here's the error message:

```
Original error message: external/fcl/include/fcl/narrowphase/detail/convexity_based_algorithm/gjk_libccd-inl.h:(1715): validateNearestFeatureOfPolytopeBeingEdge(): The origin is outside of the polytope. This should already have been identified as separating.
  Shape 1: Box0.5 0.5 0.5
  X_FS1
 0.96193976625564348026  0.15401279915737359216  0.22572537250328919556   2.0770063995786869349
-0.19134171618254486313  0.96936494951283913579  0.15401279915737359216  0.48468247475641951239
-0.19509032201612822033 -0.19134171618254486313  0.96193976625564348026   1.1543291419087275962
                      0                       0                       0                       1
  Shape 2: Box0.5 0.5 0.5
  X_FS2
 0.96193976625564348026  0.15401279915737359216  0.22572537250328919556                       2
-0.19134171618254486313  0.96936494951283913579  0.15401279915737359216                       0
-0.19509032201612822033 -0.19134171618254486313  0.96193976625564348026                    1.25
                      0                       0                       0                       1
  Solver: GjkSolver_libccd
    collision_tolerance:      2.0097183471152322134e-14
    max collision iterations: 500
    distance tolerance:       9.9999999999999995475e-07
    max distance iterations:  1000" thrown in the test body.
```

We need to reproduce this in test (see, e.g., [`test_distance_sphere_box_regression1()`](https://github.com/flexible-collision-library/fcl/blob/master/test/test_fcl_signed_distance.cpp#L467)).

Some quick observations:
  - Both boxes are the same size.
  - They have the same orientations.
  - It seems like they're stacked and canted slightly. It *seems* that they are in contact (so we would expect a small, negative signed distance), but that exact amount depends on the amount of canting. However, simply looking at the relative positions of the two transforms, its clearly on the order of 1 cm.

**Comment by SeanCurtis-TRI** [comment]
cc @hongkai-dai Looks like we've got another one. This is an interesting one because the two boxes are oriented identically but mis-aligned when stacked. If I come across anything weird, I'll let you know.

**Comment by sherm1** [comment]
Amazing! EPA has a bottomless pool of unique ways to fail.

**Comment by samzapo** [comment]
Here's another just like it
```
Original error message: external/fcl/include/fcl/narrowphase/detail/convexity_based_algorithm/gjk_libccd-inl.h:(1715): validateNearestFeatureOfPolytopeBeingEdge(): The origin is outside of the polytope. This should already have been identified as separating.
  Shape 1: Box0.25 0.25 0.25
  X_FS1
  0.95233240645725869555   0.27969320707964251405   0.12179777307008109177  0.069923301769910642389
 -0.28888687719060901493   0.95512127579560024415  0.065480689593520297054   0.23878031894890006104
-0.098017140329560589751 -0.097545161008064124042   0.99039264020161521529    1.2256137097479840037
                       0                        0                        0                        1
  Shape 2: Box0.25 0.25 0.25
  X_FS2
  0.95233240645725869555   0.27969320707964251405   0.12179777307008109177                        0
 -0.28888687719060901493   0.95512127579560024415  0.065480689593520297054                        0
-0.098017140329560589751 -0.097545161008064124042   0.99039264020161521529                     1.25
                       0                        0                        0                        1
  Solver: GjkSolver_libccd
    collision_tolerance:      2.0097183471152322134e-14
    max collision iterations: 500
    distance tolerance:       9.9999999999999995475e-07
    max distance iterations:  1000
```

**Comment by SeanCurtis-TRI** [comment]
The second case seems to be a variation on a theme. Two aligned boxes of the same size, apparently slightly in penetration. I'm going to bet that solving the first will solve the latter. Nice to have a second data point though.

**Comment by sherm1** [comment]
Wow! Awesome analysis. Solution unclear as usual with this algorithm 😢.
FYI @rpoyner-tri -- a typical example of our ongoing struggles with FCL.

**Comment by SeanCurtis-TRI** [comment]
Trying __EPA edge validation shouldn't be testing against zero__

  - This is the easiest to simply test.
  - The validation code already allows for the nearest feature actually being an edge. (When we go to expand the polytope, we use the [face normal of one of the adjacent faces](https://github.com/flexible-collision-library/fcl/blob/master/include/fcl/narrowphase/detail/convexity_based_algorithm/gjk_libccd-inl.h#L1530-L1538) as the expansion direction.) This makes the *most* sense in the co-planar case.
  - Possible issues
    - Leaking a positive signed distance out
      - Is that bad?
        - There are two operations: `nextSupport()` and `expandPolytope()` that depend on the nearest feature type.
          - `nextSupport()`
            - We compute a support direction based on a neighboring face and acquire a support vertex.
            - If the support vertex "lies on the edge", we bail out.
          - `expandPolytope()`
            - Requires that one of the edge-adjacent faces report that the new support vertex lies *outside* the face. This should be ok, because the only support vertices we allow here (from `nextSupport()`) must be measurably father from the origin than the edge. That basically guarantees that *one* of the faces will be happy with the support vertex.

The primary argument (above) against using epsilon instead of zero is about moving the boundary around which the noise affects us, but not actually handling the noise. I have a counter argument.

The GJK/EPA code should do tests relative to zero. Greater than zero is outside. Period. The code in question here is a *test*. It's validating state. The calculations it does to validate state are not necessarily the *exact* same calculations that led to the original classification. In essence, how do we define the parameters of the plane equation? Even with exactly identical vertex positions, we can still compute different normals that would be mathematically equivalent but numerically subtly different. The same applies for picking a point on the plane to solve for the displacement constant. The plane equation used by GJK is *not* available to this EPA test. So, we introduce epsilon, not to say we're moving the line of what's inside/outside by the main code, but to account for the fact that we don't actually have the values originally used. As long as the actual main code path uses zero as a hard delimiter, we keep the noise centered around zero as we would want.

Finally, after making the change, the tests pass.  :) So, it seems good until proven bad.

**Comment by hongkai-dai** [comment]
If we allow positive signed distance in the EPA algorithm, then when we call `nextSupport()`, would it choose the opposite normal direction, pointing into the polytope instead of outward from the polytope? I remember we determine the direction of the face normal based on the origin, and supposed that the origin is within the polytope.

**Comment by SeanCurtis-TRI** [comment]
You raise an interesting point, and I *do* believe it's a problem but it's a *different* problem and what you're alluding to is even *more* interesting.

In `validateNearestFeatureOfPolytopeBeingEdge()` we *compute* the normal from scratch (by calling `faceNormalPointingOutward()`). Given the face vertices we compute a normal. Think about that. We compute a normal based on edge cross products. You're suggesting we then possibly reverse the direction of that normal to confirm the signed distance to the __origin is not positive__. The *very* next thing we do with that newly computed normal is calculate the signed distance to the origin and produce a (minuscule) positive value!!!!! Clearly, the stated purpose of the method has been foiled.

In reality, in the `faceNormalPointingOutward()` method, we acknowledge the origin may be too close to trust as a witness, so, instead, we consider other vertices in the polytope to determine which side should be "outside". We're only really in trouble if both the origin and the polytope generally all lie on the same plane. Furthermore, the mechanism we use for computing distance to the origin from the plane depends on which face vertex we compute a dot product with.

I don't think introducing epsilon in the validation test changes any of that.

**Comment by hongkai-dai** [comment]
Thanks! I forgot we use the other vertices to determine the direction of the normal, when the origin is too close to the face, within a certain threshold. So supposedly if this threshold we use is consistent with the epsilon we plan to relax to, then we should be good. My only concern now is that we use a different epsilon (i.e., when we do EPA edge validation we use one epsilon, but when we compute the outward normal we use a different epsilon to determine if the origin is close to the surface, that inconsistency of epsilon could cause trouble). I suppose we use the same epsilon when we relax the EPA edge validation by epsilon, and when we compute the face normal, it is guaranteed that we won't use the origin to determine the normal direction.

**Comment by SeanCurtis-TRI** [comment]
They need not be the *same* epsilon. We just need to make sure normal computation as an epsilon *at least* as big. In this case, the epsilon is *MASSIVE* 1e-2. So, any variance that gets introduced at the 2e-16 level will be completely swallowed up.

**Comment by SeanCurtis-TRI** [comment]
Although, this *does* remind me that the fact we can't rely on the winding seems absurd. As long as we translate to simplex in a disciplined manner and expand the polytope in a disciplined manner, we can *always* know the winding at construction and not have to play these games every time we need the normal.

**Comment by SeanCurtis-TRI** [comment]
FTR The reporting user created one more scenario where the boxes were *explicitly* separated by epsilon distance and found that it also failed.

```
Original error message: external/fcl/include/fcl/narrowphase/detail/convexity_based_algorithm/gjk_libccd-inl.h:(1715): validateNearestFeatureOfPolytopeBeingEdge(): The origin is outside of the polytope. This should already have been identified as separating.
  Shape 1: Box0.25 0.25 0.25
  X_FS1
 0.94096063217417758029  0.29296840037289501035  0.16959541586174811667 0.073242100093223821977
-0.23569836841299879326  0.92661523595848427348 -0.29296840037289506586   0.2316538089896212349
 -0.2429801799032638987  0.23569836841299884878  0.94096063217417758029   1.3089245921032497399
                      0                       0                       0                       1
  Shape 2: Box0.25 0.25 0.25
  X_FS2
 0.94096063217417758029  0.29296840037289501035  0.16959541586174811667                       0
-0.23569836841299879326  0.92661523595848427348 -0.29296840037289506586                       0
 -0.2429801799032638987  0.23569836841299884878  0.94096063217417758029                    1.25
                      0                       0                       0                       1
  Solver: GjkSolver_libccd
    collision_tolerance:      2.0097183471152322134e-14
    max collision iterations: 500
    distance tolerance:       9.9999999999999995475e-07
    max distance iterations:  1000
```

This is unsurprising. But we're adding this additional case here for completeness.

**Comment by sherm1** [comment]
Should this be closed now that #494 merged?

**Comment by SeanCurtis-TRI** [comment]
It will; we haven't merged yet.
